packages/repository/src/repositories/legacy-juggler-bridge.ts

```diff
@@ -385,34 +391,67 @@ export class DefaultCrudRepository<
     }
   }
 
+  /**
+   * Template method to amend filters with foreign key sources
+   * and prune the values from result if needed
+   * @param filter - a filter to fix
+   * @param resolver - a function that will resolve with entities
+   */
+  protected async withInclusionAwareFilter(
+    filter: Filter<T> | undefined,
+    resolver: (filter?: legacy.Filter) => Promise<(T & Relations)[]>,
+  ) {
+    const include = filter?.include;
+
+    const sourceKeys = (include ?? [])
+      .map(inclusion => {
+        const inclusionResolver = this.inclusionResolvers.get(
+          inclusion.relation,
+        );
+        const meta = inclusionResolver?.relationMeta as {keyFrom: string};
+        return meta?.keyFrom as keyof T;
+      })
+      .filter(Boolean);
+
+    const {filter: fixed, pruneFields} = ensureFields(sourceKeys, filter!);
+
+    const entities = await resolver(this.normalizeFilter(fixed));
+    return entities.map(e => Object.assign(e, pruneFields));
+  }
+
   async find(
     filter?: Filter<T>,
     options?: Options,
   ): Promise<(T & Relations)[]> {
     const include = filter?.include;
-    const models = await ensurePromise(
-      this.modelClass.find(this.normalizeFilter(filter), options),
-    );
-    const entities = this.toEntities(models);
-    return this.includeRelatedModels(entities, include, options);
+    return this.withInclusionAwareFilter(filter, async legacyFilter => {
+      const models = await ensurePromise(
+        this.modelClass.find(legacyFilter, options),
+      );
+      const entities = this.toEntities(models);
+      return this.includeRelatedModels(entities, include, options);
+    });
   }
 
   async findOne(
     filter?: Filter<T>,
     options?: Options,
   ): Promise<(T & Relations) | null> {
-    const model = await ensurePromise(
-      this.modelClass.findOne(this.normalizeFilter(filter), options),
-    );
-    if (!model) return null;
-    const entity = this.toEntity(model);
     const include = filter?.include;
-    const resolved = await this.includeRelatedModels(
-      [entity],
-      include,
-      options,
+    const resolved = await this.withInclusionAwareFilter(
+      filter,
+      async legacyFilter => {
+        const model = await ensurePromise(
+          this.modelClass.findOne(legacyFilter, options),
+        );
+        if (!model) {
+          return [] as (T & Relations)[];
+        }
+        const entity = this.toEntity(model);
+        return this.includeRelatedModels([entity], include, options);
+      },
     );
-    return resolved[0];
+    return resolved[0] || null;
```

Comment:
I am not sure if we should have ` || null` here. I think the empty value is handled before retuning here?

Reply:
Not really, we can get an empty array we guard against undefined here.